### PR TITLE
Filter + Class Cards

### PR DESCRIPTION
**Pull request by JohannesSchmdit21**
Implemented Cards for the homepage

### CODE REVIEW

**Comment by JoeJYu** [comment]
Da wäre noch ein Merge conflict in dem Branch. Schau nochmal drüber

File: client/src/components/Home/Parent.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+    const url = 'http://localhost:3000/'
+
+    useEffect( () => {
+        getAllClasses();
+    }, [])
+
+    const getAllClasses = () => {
+        axios.get('${url}api/classes')
+        .then((response) => {
```
Comment (stefanwaldhauser) [inline client/src/components/Home/Parent.js]: Would be nice if we could consistencly use the more modern await async

File: client/package.json [hunk 2]
```diff
@@ -17,6 +17,7 @@
     "dayjs": "^1.10.5",
     "formik": "^2.2.8",
     "holderjs": "^2.9.9",
+    "jquery": "^2.0.0",
```
Comment (stefanwaldhauser) [inline client/package.json]: Remove? Why needed

File: client/src/components/Home/Home.js [hunk 3]
```diff
@@ -17,15 +22,22 @@ export const Home = () => {
...
+    fetchClasses(limit, skip)
+  }, [skip, limit])
 
-  const fetchClasses = async (filters) => {
+  const fetchClasses = async (limit, skip) => {
     setIsError(false);
     setIsLoading(true);
     try {
-      const result = await axios('/api/classes');
+      const result = await axios(`/api/classes?limit=${limit}&skip=${skip}${queryString}`);
```
Comment (stefanwaldhauser) [inline client/src/components/Home/Home.js]: https://masteringjs.io/tutorials/axios/get-query-params 
This works but please look at the way how it is recommended to pass query parameters via axios

Comment (stefanwaldhauser) [reply client/src/components/Home/Home.js]: You coud simply store all the set filters in a state for example

File: client/src/components/Home/Home.js [hunk 4]
```diff
@@ -95,22 +116,44 @@ export const Home = () => {
...
             </Col>
           </Row>
           <p></p>
           <Row> {classCards} </Row>
+          <label htmlFor="">
+      </label>
+      
+				
+          <div>
+            <Button onClick={previousPage}>Previos Page</Button>
```
Comment (stefanwaldhauser) [inline client/src/components/Home/Home.js]:
```suggestion
            <Button onClick={previousPage}>Previous Page</Button>
```

File: client/src/components/Home/Home.js [hunk 5]
```diff
@@ -35,24 +47,33 @@ export const Home = () => {
...
-      return data.meetingAddress.city.search(value) != -1;
-    });
-    setFilteredClasses(result);
+    setQueryString(queryString.concat(`&city=${value}`));
+    setCity(value);
   };
 
+    const handleFilterDate = (e) => {
+    setQueryString(queryString.concat(`&date=${e}`));
+    localStorage.setItem("date", e.target.value)
```
Comment (stefanwaldhauser) [inline client/src/components/Home/Home.js]: Why are you using the local storage here? What is the useCase?

File: server/src/models/index.js [hunk 6]
```diff
@@ -0,0 +1,13 @@
+// const dbConfig = require('../.env.example');
```
Comment (stefanwaldhauser) [inline server/src/models/index.js]: I am a little bit confused what that file is doing exactly. Lets discuss this today in the meeting.

File: server/src/routes/api/classes.js [hunk 7]
```diff
@@ -17,29 +17,48 @@ var photosUpload = upload.fields([
...
+
+    const { hostId, category, city, date } = req.query;
     let classes = [];
+
     if (hostId) {
       // classes by the specified host
-      classes = await Class.find({ host: new ObjectId(hostId) }).populate('host');
+      classes = await Class.find({host: new ObjectId(hostId)}).populate('host').skip(skipValue).limit(limitValue)
+    }  
+    else if (category) {
```
Comment (stefanwaldhauser) [inline server/src/routes/api/classes.js]: If I understand this code directly I can no filter both for category and dates? This seems broken kind of. A user wants to have multiple filter active at the same time.

**Review by stefanwaldhauser — CHANGES_REQUESTED** [review]
Some Change Requests

File: server/src/routes/api/classes.js [hunk 8]
```diff
@@ -125,11 +151,16 @@ router.put('/:id', [requireJwtAuth, photosUpload], async (req, res, next) => {
...
+    if (err.message) {
+      res.status(500).json({ message: err.message });
+    } else {
+      res.status(500).json({ message: 'Something went wrong during the class creation.' });
+    }
   }
 });
 
 router.post('/', [requireJwtAuth, photosUpload], async (req, res, next) => {
+  debugger;
```
Comment (stefanwaldhauser) [inline server/src/routes/api/classes.js]: Please remove debugger

File: server/src/routes/api/classes.js [hunk 9]
```diff
@@ -17,52 +17,78 @@ var photosUpload = upload.fields([
...
+      res.status(500).json({ message: err.message });
+    } else {
+      res.status(500).json({ message: 'Something went wrong during the class creation.' });
+    }
   }
 });
 
-router.delete('/:id', [requireJwtAuth], async (req, res, next) => {
+router.delete('/:id', requireJwtAuth, async (req, res) => {
+  debugger;
```
Comment (stefanwaldhauser) [inline server/src/routes/api/classes.js]: Another debugger code.....

File: client/src/components/Users/Users old.js [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+                          <span className="info">
+                            {moment(user.createdAt).format('dddd, MMMM Do YYYY, H:mm:ss')}
+                          </span>
+                        </div>
+                      </Card.Text>
+                    </Card.Body>
+                  </Card>
+
+                );
+                //mapping ende
```
Comment (stefanwaldhauser) [inline client/src/components/Users/Users old.js]: What does this mean?

File: client/src/components/Users/HostCard.js [hunk 11]
```diff
@@ -0,0 +1,57 @@
+import React from 'react';
+import { Button, Card, ListGroup } from 'react-bootstrap';
+import { run as runHolder } from 'holderjs/holder';
+import { useEffect } from 'react';
+import dayjs from 'dayjs';
+import utc from 'dayjs/plugin/utc';
+import localizedFormat from 'dayjs/plugin/localizedFormat';
+import { Link } from 'react-router-dom';
+import moment from 'moment';
```
Comment (stefanwaldhauser) [inline client/src/components/Users/HostCard.js:9]: We are not using moment. It is just a leftover from the boilerplate. Please remove it in the whole code and in the package.json and replace with dayjs

File: client/src/components/Users/Users old.js [hunk 12]
```diff
@@ -0,0 +1,93 @@
+import React, { useEffect } from 'react';
+import { compose } from 'redux';
+import { connect } from 'react-redux';
+import { Link } from 'react-router-dom';
+import moment from 'moment';
```
Comment (stefanwaldhauser) [inline client/src/components/Users/Users old.js]: Remove moment and switch to dayjs

File: server/src/models/index.js [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+
+const mongoose = require('mongoose');
+const mongoosePaginate = require('mongoose-paginate-v2');
+
+mongoose.Promise = global.Promise;
```
Comment (stefanwaldhauser) [inline server/src/models/index.js]: quick googling shows that this is really legacy code.https://stackoverflow.com/questions/51862570/mongoose-why-we-make-mongoose-promise-global-promise-when-setting-a-mongoo

Remove all this unneeded stuff

File: server/src/routes/api/classes.js [hunk 14]
```diff
@@ -17,52 +17,78 @@ var photosUpload = upload.fields([
 
 router.get('/', async (req, res, next) => {
   try {
-    const { hostId } = req.query;
+    const limitValue = parseInt(req.query.limit);
```
Comment (stefanwaldhauser) [inline server/src/routes/api/classes.js:90]: You are not checking if this query option is even sent? What happens if limit is not sent?

File: client/src/components/Home/Home.js [hunk 15]
```diff
@@ -57,9 +114,46 @@ export const Home = () => {
...
+
+             <Button className="mx-auto" onClick={fetchClasses}>Submit</Button>
+             </form>
+             <Button onClick={fetchClasses}>Show all</Button>
+
+            </Col>
+          </Row>
+          <p></p>
+          <Row> {classCards} </Row>
+          <label htmlFor="">
```
Comment (stefanwaldhauser) [inline client/src/components/Home/Home.js]: What is this?

File: client/src/components/Home/Filters.js [hunk 16]
```diff
@@ -0,0 +1,30 @@
+//ToDo: Add label to the left side of the filters
+//ToDo: Add search icon button to search with filters --> Collect filter and only if the user presses search apply filter
+//ToDo: Make it possible to clear a filter
```
Comment (stefanwaldhauser) [inline client/src/components/Home/Filters.js:3]: All of this was the goal for this sprint? Please contact the team if you are sure you will not make it...

File: client/src/components/Home/Filters.js [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+
+import React from 'react';
+import 'bootstrap/dist/js/bootstrap.bundle';
```
Comment (stefanwaldhauser) [inline client/src/components/Home/Filters.js]: Why are you importing bootstrap her?

File: client/src/components/Home/Filters.js [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+      <select
+        className="form-control selectpicker"
+        onChange={fun}
+        id="select-criteria"
+        data-live-search="true"
+      >
+        <option value="" disabled selected>
+          {prompt}
+        </option>
+        for (let i of options){' '}
```
Comment (stefanwaldhauser) [inline client/src/components/Home/Filters.js]: This seems to be just a string?

File: client/src/components/Home/Parent.js (hunk 1 above)
Comment (chrisdeu) [reply client/src/components/Home/Parent.js]: Removed the whole component

File: client/package.json (hunk 2 above)
Comment (chrisdeu) [reply client/package.json]: Removed

File: client/src/components/Home/Home.js (hunk 4 above)
Comment (chrisdeu) [reply client/src/components/Home/Home.js]: No American English?

File: client/src/components/Home/Home.js (hunk 5 above)
Comment (chrisdeu) [reply client/src/components/Home/Home.js]: Removed, was left from first implementation version

File: server/src/models/index.js (hunk 6 above)
Comment (chrisdeu) [reply server/src/models/index.js]: Removed, was left from another implementation version

File: server/src/routes/api/classes.js (hunk 8 above)
Comment (chrisdeu) [reply server/src/routes/api/classes.js]: Removed

File: server/src/routes/api/classes.js (hunk 9 above)
Comment (chrisdeu) [reply server/src/routes/api/classes.js]: Removed

File: server/src/models/index.js (hunk 13 above)
Comment (chrisdeu) [reply server/src/models/index.js]: Removed

File: server/src/routes/api/classes.js (hunk 14 above)
Comment (chrisdeu) [reply server/src/routes/api/classes.js:90]: The assumption is that the client always sends a value but a check would make sense

File: client/src/components/Home/Filters.js (hunk 16 above)
Comment (chrisdeu) [reply client/src/components/Home/Filters.js:3]: Did not notice the todos, sorry. Will address the design-related requests till  27.06

File: client/src/components/Home/Filters.js (hunk 17 above)
Comment (chrisdeu) [reply client/src/components/Home/Filters.js]: Removed, left from a previos implementation version

File: client/src/components/Home/Filters.js (hunk 18 above)
Comment (chrisdeu) [reply client/src/components/Home/Filters.js]: Removed

File: server/src/routes/api/classes.js [hunk 19]
```diff
@@ -15,25 +15,102 @@ var photosUpload = upload.fields([
...
+  let { city, category, date, guests, price, rating, limit, skip } = req.body;
+
+  if (date) {
+    date = dayjs(date);
+  }
+
+  try {
+    let classes = await Class.find();
+    // Apply City Filter
+    if (city) {
```
Comment (stefanwaldhauser) [inline server/src/routes/api/classes.js:32]: Very good code :)

**Pull request merged**